spinalcordtoolbox/centerline/core.py

```diff
@@ -175,9 +176,30 @@ def get_centerline(im_seg, param=ParamCenterline(), verbose=1, remove_temp_files
         im_centerline.data = np.zeros(im_centerline.data.shape)
         # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
         # TODO: check this round and clip-- suspicious
-        im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
-                           round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
-                           z_ref] = 1
+        # Binary
+        if param.soft == 0:
+            im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
+                               round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
+                               z_ref] = 1
+        # Soft
+        else:
+            im_centerline.data[np.ceil(x_centerline_fit).astype(int),
+                               np.ceil(y_centerline_fit).astype(int),
+                               z_ref] = 0.5 * ((x_centerline_fit - x_centerline_fit.astype(int)) +
+                                               (y_centerline_fit - y_centerline_fit.astype(int)))
+            im_centerline.data[np.floor(x_centerline_fit).astype(int),
+                               np.floor(y_centerline_fit).astype(int),
+                               z_ref] = 0.5 * (1 - (x_centerline_fit - x_centerline_fit.astype(int)) +
+                                               (1 - (y_centerline_fit - y_centerline_fit.astype(int))))
+            im_centerline.data[np.floor(x_centerline_fit).astype(int),
+                               np.ceil(y_centerline_fit).astype(int),
+                               z_ref] = 0.5 * (1 - (x_centerline_fit - x_centerline_fit.astype(int)) +
+                                               (y_centerline_fit - y_centerline_fit.astype(int)))
+            im_centerline.data[np.ceil(x_centerline_fit).astype(int),
+                               np.floor(y_centerline_fit).astype(int),
+                               z_ref] = 0.5 * (x_centerline_fit - x_centerline_fit.astype(int) +
+                                               (1 - (y_centerline_fit - y_centerline_fit.astype(int))))
```

Comment:
I think we could maybe save some space, and make things a little more readable, by precomputing some of the repeated quantities?

```python
            # Compute integer pixel locations (all together, these form a 2x2 grid around centerline)
            x_ceil = np.ceil(x_centerline_fit).astype(int)
            y_ceil = np.ceil(y_centerline_fit).astype(int)
            x_floor = np.floor(x_centerline_fit).astype(int)
            y_floor = np.floor(y_centerline_fit).astype(int),
            # Compute the difference between the centerline and its rounded values
            x_diff = x_centerline_fit - x_centerline_fit.astype(int)
            y_diff = y_centerline_fit - y_centerline_fit.astype(int)
            # Assign soft segmentation values to the 2x2 grid based on the computed differences
            im_centerline.data[x_ceil, y_ceil, z_ref]   = 0.5 * (x_diff       + y_diff)
            im_centerline.data[x_floor, y_floor, z_ref] = 0.5 * ((1 - x_diff) + (1 - y_diff))
            im_centerline.data[x_floor, y_ceil, z_ref]  = 0.5 * ((1 - x_diff) + y_diff))
            im_centerline.data[x_ceil, y_floor, z_ref]  = 0.5 * (x_diff       + (1 - y_diff))
```

I think this makes it easier to intuitively see how the soft segmentation is computed.

Reply:
Maybe we could say like, `x__ceil` or something, and use extra brackets, as a way to better align the calls? Hehe...

```python
      # Assign soft segmentation values to the 2x2 grid based on the computed differences
      im_centerline.data[x__ceil, y__ceil, z_ref] = 0.5 * ((x_diff)     + (y_diff))
      im_centerline.data[x_floor, y_floor, z_ref] = 0.5 * ((1 - x_diff) + (1 - y_diff))
      im_centerline.data[x_floor, y__ceil, z_ref] = 0.5 * ((1 - x_diff) + (y_diff))
      im_centerline.data[x__ceil, y_floor, z_ref] = 0.5 * ((x_diff)     + (1 - y_diff))
```

Probably not too important, but I really like when everything aligns all nicely. :stuck_out_tongue_closed_eyes:

Reply:
Awesome suggestion! 👍🏻  Applied in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026/commits/5ec33d6416c936ebe2af657224f311ee13a2aa93